code/game/mecha/equipment/tools/janitor_tools.dm

```diff
@@ -0,0 +1,363 @@
+	var/turf/T2 = get_step(T, turn(direction, -90))
+	var/list/the_targets = list(T, T1, T2)
+	playsound(chassis, 'sound/effects/spray2.ogg', 75, TRUE, -3)
+	for(var/turf/target_turf in the_targets)
+		INVOKE_ASYNC(src, PROC_REF(spray), target_turf)
+
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_spray/proc/spray(turf/target)
+	var/obj/effect/decal/chempuff/spray = new /obj/effect/decal/chempuff(get_turf(chassis))
+	var/datum/reagents/R = new/datum/reagents(5)
+	spray.reagents = R
+	R.my_atom = spray
+	reagents.trans_to(spray, 5)
+	spray.icon += mix_color_from_reagents(spray.reagents.reagent_list)
+	for(var/B in 1 to 4)
+		if(QDELETED(spray))
+			return
+		if(spray.reagents.total_volume == 0)
+			qdel(spray)
+			return
+		step_towards(spray, target)
+		if(QDELETED(spray))
+			return
+		var/turf/spray_turf = get_turf(spray)
+		spray.reagents.reaction(spray_turf)
+		for(var/atom/atm in spray_turf)
+			spray.reagents.reaction(atm)
+		if(spray.loc == target)
+			qdel(spray)
+			return
+		sleep(2)
+	qdel(spray)
+
+// Auto-regeneration of space cleaner. Takes energy.
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_spray/process()
+	if(reagents.total_volume < 100)
+		reagents.add_reagent(refill_reagent, refill_rate)
+		chassis.use_power(refill_cost)
+		update_equip_info()
+
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_spray/get_equip_info()
+	var/output = ..()
+	if(output)
+		return "[output] \[<a href='byond://?src=[UID()];toggle_mode=1'>Refill [refill_enabled? "Enabled" : "Disabled"]</a>\] \[[reagents.total_volume]\]"
+
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_spray/Topic(href,href_list)
+	..()
+	var/datum/topic_input/afilter = new (href,href_list)
+	if(afilter.get("toggle_mode"))
+		refill_enabled = !refill_enabled
+		if(refill_enabled)
+			START_PROCESSING(SSobj, src)
+		else
+			STOP_PROCESSING(SSobj, src)
+		update_equip_info()
+
+// Garbage Magnet
+/obj/item/mecha_parts/mecha_equipment/janitor/garbage_magnet
+	name = "\improper WA1E Garbage Magnet"
+	desc = "Bluespace technology integrated with an oversized garbage bag and heavy duty magnets allows this device to pick up all manner of litter. \
+	The complex technology prevents users from directly looking inside the bag."
+	icon_state = "mecha_trash_magnet"
+	equip_cooldown = 1.5 SECONDS
+	energy_drain = 5
+	range = MECHA_MELEE | MECHA_RANGED
+	/// Toggle for filling the bag (true) or emptying (false)
+	var/bagging = TRUE
+	/// Toggle for wide area or single tile pickups
+	var/extended = FALSE
+	/// List of items currently in the bag
+	var/list/cargo = list()
+	/// How many different items can be in the bag?
+	var/cargo_slots = 100
+	/// How much weight is the maximum for the bag?
+	var/cargo_max_weight = 100
+	/// Largest weight class that can fit in the bag
+	var/max_weight_class = WEIGHT_CLASS_NORMAL
+	/// List of items the bag cannot hold
+	var/cant_hold = list(/obj/item/disk/nuclear, /obj/item/grown/bananapeel/traitorpeel, /obj/item/storage/bag)
```

Comment:
Can I ask why traitorpeels are specifically excluded here?

I think you also need to deliberately make this a typecache yourself, probably somewhere in init or new. I don't remember the function name, but that will populate the list with all of the subtypes as needed.

Reply:
Traitorpeels are excluded as regular trash bags, as well as bluespace trash bags, also exclude them. They were kept excluded for consistency between trash bags.

Reply:
Typecache has been added to the Initialize function.